### DISCUSSION TITLE
Spacy ValueError: could not broadcast input array from shape (1258,64) into shape (1226,64)

### DISCUSSION THREAD

**Discussion by smisra1** [explosion/spaCy#8864] [Help: Coding & Implementations] [answered]
My spacy version is 2.3.7.

The aim is to retrain my existing custom NER model multiple times with incoming batches of input data. Overlooking "catastrophic forgetting", I still want to retrain the model multiple times.

I could run, load and retrain my custom model 2 times successfully. But, on trying to run the model the 3rd time, I get the error for the code `nlp = spacy.load('mymodel')` as:

`ValueError: could not broadcast input array from shape (1258,64) into shape (1226,64)`

What could be the source of the problem and how do I rectify it?

**Comment by svlandeg** [comment, marked as answer]
Hi!

Unfortunately that's a very generic error message. It typically points to some internal parameters not being set correctly (anymore).

I can't really provide more detailed pointers without more specific information... Could you create a minimal code snippet that reproduces the error? (e.g. creating a new model, training on just a few data points, saving out, loading back in, training again).

Alternatively, can you share the full error stack trace? Is the error reproducible (i.e. does it always occur when running the same code snippet)? What changed in between running the training step the first two times, and the third one?

**Reply by smisra1** [reply]
Hi!
The full error message is-
```
Traceback (most recent call last):
  File "C:\Users\smisr\PycharmProjects\june_2021\annotator_nlp\update_api.py", line 7, in <module>
    nlp = spacy.load("update2")
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\__init__.py", line 30, in load
    return util.load_model(name, **overrides)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\util.py", line 172, in load_model
    return load_model_from_path(Path(name), **overrides)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\util.py", line 222, in load_model_from_path
    return nlp.from_disk(model_path, exclude=disable)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\language.py", line 974, in from_disk
    util.from_disk(path, deserializers, exclude)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\util.py", line 690, in from_disk
    reader(path / key)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\language.py", line 968, in <lambda>
    deserializers[name] = lambda p, proc=proc: proc.from_disk(
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\pipeline\entityruler.py", line 358, in from_disk
    from_disk(path, deserializers_patterns, {})
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\util.py", line 690, in from_disk
    reader(path / key)
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\pipeline\entityruler.py", line 344, in <lambda>
    "patterns": lambda p: self.add_patterns(
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\pipeline\entityruler.py", line 226, in add_patterns
    for label, pattern, ent_id in zip(
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\language.py", line 829, in pipe
    for doc in docs:
  File "C:\Users\smisr\PycharmProjects\june_2021\virtualenv\lib\site-packages\spacy\language.py", line 1149, in _pipe
    for doc in docs:
  File "nn_parser.pyx", line 254, in pipe
  File "nn_parser.pyx", line 274, in spacy.syntax.nn_parser.Parser.predict
  File "nn_parser.pyx", line 286, in spacy.syntax.nn_parser.Parser.greedy_parse
  File "nn_parser.pyx", line 197, in spacy.syntax.nn_parser.Parser._resize
  File "_parser_model.pyx", line 269, in spacy.syntax._parser_model.ParserModel.resize_output
ValueError: could not broadcast input array from shape (1258,64) into shape (1226,64)

```
My model pipeline contains multiple entity rulers and ner pipes which are to be retrained.
I can add the main codes for retraining the different parts of the model pipeline, if that would be helpful.

Below is the main code I used to retrain my entity ruler pipes in the model:
```
            def update_er(e_name, pattern):

                ruler = nlp.get_pipe(e_name)
                for i in pattern:
                    ruler.add_patterns(i)


            update_er('entity_ruler', new_dict)

```
Below is the main code I used to retrain the ner pipes in the model:

```
            def train_spacy(data, p_name, iterations):
                TRAIN_DATA = data
                # add labels
                ner = nlp.get_pipe(p_name)
                for _, annotations in TRAIN_DATA:
                    for ent in annotations.get('entities'):
                        ner.add_label(ent[2])

                other_pipes = [pipe for pipe in nlp.pipe_names if pipe != p_name]
                with nlp.disable_pipes(*other_pipes):

                    optimizer = nlp.resume_training()
                    for itn in range(iterations):
                        print('Starting iteration ' + str(itn))

                        losses = {}
                        batches = minibatch(TRAIN_DATA, size=64)
                        for batch in batches:

                            text, annotation = zip(*batch)
                            try:
                                nlp.update(
                                    text,
                                    annotation,
                                    sgd=optimizer,
                                    drop=0.3,
                                )
                            except Exception as e:
                                pass

train_spacy(new_data, 'ner', 20)

```

Yes it always occurs when I run it a 3rd try. There was no difference. Just to test if retraining was occurring, I used around 10 examples for each try.
Plan of action for every try was-
I run the model once, retrain it with 10 examples, test it, save it using `nlp.to_disk()`, stop running it and repeat.

Meaning I could successfully retrain the model once and load it back. But when I tried to retrain the updated model, it did get retrained but would not load again after, giving the above error.

**Reply by svlandeg** [reply]
Are you training it each time with the exact same examples? 

From the stack trace, it looks like the issue occurs when the parser tries to resize itself, which is what happens automatically when it encounters a label in the gold data that it hadn't seen before. That means that it's either a data problem (a new label suddenly showing up?) or an initialization/serialization problem.

Do the entity rulers introduce NER labels that are not part of the NER model? Could you double check whether those labels are present in your `ner` component after training the first time, after loading the model back in, etc? I'm suspecting some kind of error in the parser's internal state when serializing, which could ultimately cause this.

**Reply by smisra1** [reply]
The examples are different in each try for retraining it. Their labels are known to the model during the original training and are not new to it. There might come a time in the future when a new label would be introduced but that would be later on.
For now, no new labels were introduced.
Is there any way to resolve this issue?

**Reply by smisra1** [reply]
Hi. Looking forward to your reply. I really need a solution for this. I am on a deadline.

**Reply by svlandeg** [reply]
I wasn't clear on the answer to my earlier question:
> Do the entity rulers introduce NER labels that are not part of the NER model? 

Can you give an overview of which exact components are in your pipeline, in which order, and which entity labels each of them introduce (with a focus on the NER and entity rulers)

**Reply by svlandeg** [reply]
I also hadn't seen a detailed response to 
> Could you double check whether those labels are present in your ner component after training the first time, after loading the model back in, etc?

Also,

> I really need a solution for this. I am on a deadline.

I'm sorry to say that we don't provide any warranty/SLA for the free support we provide for this free open-source software ;-) That said, I'm happy to try and help, but without an immediate reproducible code snippet, I can't help debug, and I need some additional information from you to try and find the culprit.

**Reply by smisra1** [reply]
> I wasn't clear on the answer to my earlier question:
> 
> > Do the entity rulers introduce NER labels that are not part of the NER model?
> 
> Can you give an overview of which exact components are in your pipeline, in which order, and which entity labels each of them introduce (with a focus on the NER and entity rulers)

My pipeline is in this order-
-entity ruler a
-ner a
-entity ruler b
-ner b
-entity ruler c
-ner c
-entity ruler d
-ner d

There are around multiple different custom entity labels for each ner that they are trained to work on. Each ner has been trained on its own custom data separately.
The entity labels are for my organisation's related work. Example - label names for ner a would be, 'Message', 'Damage', etc
While testing, I only used a particular entity ruler and ner pipes together for my output.

**Reply by smisra1** [reply]
> I also hadn't seen a detailed response to
> 
> > Could you double check whether those labels are present in your ner component after training the first time, after loading the model back in, etc?
> 
> Also,
> 
> > I really need a solution for this. I am on a deadline.
> 
> I'm sorry to say that we don't provide any warranty/SLA for the free support we provide for this free open-source software ;-) That said, I'm happy to try and help, but without an immediate reproducible code snippet, I can't help debug, and I need some additional information from you to try and find the culprit.

Yes, more labels were added in the ner component when I checked the meta.json file for the updated model and the original model. It was more after each updation.

Sorry for not being able to provide a reproducible code snippet. Since, the model has data that cannot be disclosed without my organisation's consent, I am unable to display the code without giving out the data.

**Reply by svlandeg** [reply]
Ok so it looks like this is definitely an interaction between the labels of the entity rulers, and the labels of the NER models. Some background on a previous, related bug can be found here: https://github.com/explosion/spaCy/issues/4042#issuecomment-518419990 and https://github.com/explosion/spaCy/pull/4330

Could you try the following: for each NER, not only add the labels of the training data for that NER, but also ALL labels of all entity rulers that run before that specific NER? Add those labels early on when defining the pipeline, and only then start training & serialization. Hopefully the IO shouldn't come into an inconsistent state then.

**Reply by smisra1** [reply]
I got the same error after trying it out. The difference was in the shape written in the error message:

Earlier  error-
`ValueError: could not broadcast input array from shape (1258,64) into shape (1226,64)`

New error message-
`ValueError: could not broadcast input array from shape (1314,64) into shape (1282,64)`

I noticed a few more changes on updated models in the cfg file for each ner-
-`nr_class increased with every updation`
-`deprecation_fixes: {"vectors_name": "en_model.vectors"}` was added to the ner cfg files in the updated models

**Reply by svlandeg** [reply]
If `nr_class` keeps increasing, it would be good to inspect which exact `labels` are in each NER model at various stages - it looks like the NER models are still expanding "on the fly", i.e. they don't know all the possible entity labels yet beforehand?

**Reply by smisra1** [reply]
> Ok so it looks like this is definitely an interaction between the labels of the entity rulers, and the labels of the NER models. Some background on a previous, related bug can be found here: [#4042 (comment)](https://github.com/explosion/spaCy/issues/4042#issuecomment-518419990) and #4330
> 
> Could you try the following: for each NER, not only add the labels of the training data for that NER, but also ALL labels of all entity rulers that run before that specific NER? Add those labels early on when defining the pipeline, and only then start training & serialization. Hopefully the IO shouldn't come into an inconsistent state then.

I think I might have misunderstood your suggestion. Is this task to be done before training the original model i.e. from scratch? Meaning should I start fresh, add the labels when I first created the pipes and then save this model and use it to be updated further?

Because I only added the labels of the entity rulers using `nlp.get_pipe()`  and  `ner.add_label()`method to an existing, trained custom pipeline of the order I mentioned before.

**Reply by svlandeg** [reply]
Yes, that was my suggestion:
> Add those labels early on when defining the pipeline, and only then start training & serialization.

**Reply by smisra1** [reply]
It worked! Thank you so much! 
I have just 2 more doubts. 

1. On updating the same trained model, do I need to add the new training data labels of the entity rulers to the ner pipes every time?
2. Will the model encounter an error if, in the future, a new label that the whole model has never come across, is added into the new training data?

**Reply by svlandeg** [reply]
Happy to hear we found a workaround that lets you continue with your work, in any case.

At the same time this does feed my suspicion that this is a bug in spaCy v2. The intention is definitely that the NER can resize itself and this should work also after training & serialization. There seems to be a complex interaction here with the entity rulers though. The thing is that each entity ruler also runs part of the pipeline upon initialization, which means they run the NER models, which in turn are influenced by the labels predicted by the entity rulers.

This might be fixed in spaCy v3, though I'm not 100% sure. And to be able to fully debug this, we'd need to come up with a reproducible code snippet first.

> On updating the same trained model, do I need to add the new training data labels of the entity rulers to the ner pipes every time?

If you added those labels before training, they should be serialized to disk and you shouldn't need to add them again

> Will the model encounter an error if, in the future, a new label that the whole model has never come across, is added into the new training data?

That shouldn't happen, no, but as long as we're not 100% certain what exactly is causing the issue you've been seeing, I can't say for sure.

**Reply by smisra1** [reply]
Noted. I tried to create dummy data for a reproducible code snippet but the same error couldn't be captured. Thank you for helping. I will be sure to add the reproducible code once I am able to get the same error.

**Reply by svlandeg** [reply]
Great! I've also created a bug ticket on the issue tracker: https://github.com/explosion/spaCy/issues/8923
